test/unit/nonce-tracker-test.js

```diff
@@ -12,29 +14,65 @@ describe('Nonce Tracker', function () {
         'from': '0x7d3517b0d011698406d6e0aed8453f0be2697926',
         'gas': '0x30d40',
         'value': '0x0',
+        'nonce': '0x3',
+      },
+    }]
+    confirmedTxs = [{
+      'status': 'confirmed',
+      'txParams': {
+        'from': '0x7d3517b0d011698406d6e0aed8453f0be2697926',
+        'gas': '0x30d40',
+        'value': '0x0',
         'nonce': '0x0',
       },
+    }, {
+      'status': 'confirmed',
+      'txParams': {
+        'from': '0x7d3517b0d011698406d6e0aed8453f0be2697926',
+        'gas': '0x30d40',
+        'value': '0x0',
+        'nonce': '0x1',
+      },
+    }, {
+      'status': 'confirmed',
+      'txParams': {
+        'from': '0x7d3517b0d011698406d6e0aed8453f0be2697926',
+        'gas': '0x30d40',
+        'value': '0x0',
+        'nonce': '0x2',
+      },
     }]
 
 
     getPendingTransactions = () => pendingTxs
+    getConfirmedTransactions = () => confirmedTxs
+    providerResultStub.result = '0x3'
     provider = {
-      sendAsync: (_, cb) => { cb(undefined, {result: '0x0'}) },
+      sendAsync: (_, cb) => { cb(undefined, providerResultStub) },
       _blockTracker: {
         getCurrentBlock: () => '0x11b568',
       },
     }
     nonceTracker = new NonceTracker({
       provider,
       getPendingTransactions,
+      getConfirmedTransactions,
     })
   })
 
   describe('#getNonceLock', function () {
     it('should work', async function () {
       this.timeout(15000)
       const nonceLock = await nonceTracker.getNonceLock('0x7d3517b0d011698406d6e0aed8453f0be2697926')
-      assert.equal(nonceLock.nextNonce, '1', 'nonce should be 1')
+      assert.equal(nonceLock.nextNonce, '4', 'nonce should be 4')
+      await nonceLock.releaseLock()
+    })
+
+    it('should use localNonce if network returns a nonce lower then a confirmed tx in state', async function () {
```

Comment:
Additional assertions that could be valuable:

- `should generate the next nonce number even if multiple pending txs have same nonce`
- `should return 0 if there are no previous transactions`
- `should return one over network nonce if no local nonces are higher`
- `should return one over pending nonce if no other nonces are higher`